subprojects/docs/src/docs/userguide/javaPlugin.adoc

```diff
@@ -520,11 +525,18 @@ You can then register it for incremental compilation using the `getSupportedOpti
 </sample>
 ++++
 
-Processors that don't fit these categories will result in full recompilation.
-This includes processors that use `java.io` instead of the `Filer` API and processors that need to read or write resource files.
+Both categories have the following limitations:
+
+* Must not have any side effects except for using the `Filer` and `Messager` APIs.
+For instance, writing files using `java.io` will result in silent failures later on, as these files won't be cleaned up correctly.
+If your processor does this, it cannot be incremental.
+* Must not depend on compiler-specific APIs like `com.sun.source.util.Trees`. Gradle wraps the processing APIs, so attempts to cast to compiler-specific types will fail.
+If your processor does this, it cannot be incremental, unless you have some fallback mechanism.
+* Using `Filer#getResource` currently results in full recompilation. See https://github.com/gradle/gradle/issues/4701[gradle/issues/4701]
+* Using `Filer#createResource` currently results in full recompilation. See https://github.com/gradle/gradle/issues/4702[gradle/issues/4702]
 
 ====== Isolating annotation processors
-These look at each annotated element in isolation, creating generated files or validation messages for it.
+The most efficient category, these have to look at each annotated element in isolation, creating generated files or validation messages for it.
```

Comment:
Efficient in what sense?

Reply:
Efficient because we don't need to recompile every annotated file (like for aggregating processors), but only the ones that are affected by the change.

Reply:
So efficient time-wise? Memory-wise? Computationally? All of these? Perhaps something like:

    The most efficient category in terms of build times, ...

That's probably what most folks are interested in. It would be fine to add other metrics as well, like "build times and CPU usage" if that makes sense to you.